File: export-schema.php
<?php
/**
 * MySQL Schema Export Utility
 * 
 * Exports MySQL/MariaDB database schema to JSON format with comprehensive metadata
 * including columns, constraints, indexes, foreign keys, and DDL statements.
 *
 * @author Bill Nobes
 * @license MIT
 * @version 1.0.0
 */

// Parse command line arguments
function parseArgs($argv) {
    $options = [
        'host' => null,
        'database' => null,
        'user' => null,
        'password' => null,
        'port' => null,
        'output' => null,
        'filter' => null,
        'help' => false
    ];
    
    for ($i = 1; $i < count($argv); $i++) {
        $arg = $argv[$i];
        
        if ($arg === '--help' || $arg === '-h') {
            $options['help'] = true;
        } elseif (strpos($arg, '--') === 0) {
            $key = substr($arg, 2);
            $value = isset($argv[$i + 1]) ? $argv[$i + 1] : true;
            
            switch ($key) {
                case 'host':
                    $options['host'] = $value;
                    $i++;
                    break;
                case 'database':
                case 'db':
                    $options['database'] = $value;
                    $i++;
                    break;
                case 'user':
                    $options['user'] = $value;
                    $i++;
                    break;
                case 'password':
                    $options['password'] = $value;
                    $i++;
                    break;
                case 'port':
                    $options['port'] = $value;
                    $i++;
                    break;
                case 'output':
                    $options['output'] = $value;
                    $i++;
                    break;
                case 'filter':
                    $options['filter'] = $value;
                    $i++;
                    break;
            }
        }
    }
    
    return $options;
}

function showHelp() {
    echo "MySQL Schema Export Utility\n\n";
    echo "Usage: php export-schema.php [OPTIONS]\n\n";
    echo "Options:\n";
    echo "  --host HOST          Database host (default: localhost)\n";
    echo "  --database DB        Database name (required)\n";
    echo "  --user USER          Database user (required)\n";
    echo "  --password PASS      Database password\n";
    echo "  --port PORT          Database port (default: 3306)\n";
    echo "  --output DIR         Output directory (default: ./export)\n";
    echo "  --filter REGEX       Table name filter regex (default: /.*/ for all tables)\n";
    echo "  --help, -h           Show this help message\n\n";
    echo "Environment Variables:\n";
    echo "  DB_HOST, DB_NAME, DB_USER, DB_PASS, DB_PORT\n";
    echo "  OUTPUT_DIR, TABLE_NAME_REGEXP, SCHEMA_VERSION\n\n";
    echo "Examples:\n";
    echo "  php export-schema.php --database mydb --user root\n";
    echo "  php export-schema.php --db mydb --user root --filter '/user_.*/'\n";
    echo "  DB_NAME=mydb DB_USER=root php export-schema.php\n";
}

// Load configuration from config.ini if it exists
function loadConfigFile($filename = 'config.ini') {
    if (!file_exists($filename)) {
        return [];
    }
    
    $config = parse_ini_file($filename, true);
    if ($config === false) {
        echo "Warning: Could not parse config file: $filename" . PHP_EOL;
        return [];
    }
    
    return $config;
}

$configFile = loadConfigFile();
$cliArgs = parseArgs($argv ?? []);

if ($cliArgs['help']) {
    showHelp();
    exit(0);
}

// === CONFIGURATION ===
// Database connection settings - priority: CLI args > environment variables > config file > defaults
$DB_HOST = $cliArgs['host'] ?: getenv('DB_HOST') ?: ($configFile['database']['host'] ?? 'localhost');
$DB_NAME = $cliArgs['database'] ?: getenv('DB_NAME') ?: ($configFile['database']['name'] ?? '');
$DB_USER = $cliArgs['user'] ?: getenv('DB_USER') ?: ($configFile['database']['user'] ?? '');
$DB_PASS = $cliArgs['password'] ?: getenv('DB_PASS') ?: ($configFile['database']['password'] ?? '');
$DB_PORT = $cliArgs['port'] ?: getenv('DB_PORT') ?: ($configFile['database']['port'] ?? 3306);

// Table name filter (regex pattern)
$TABLE_NAME_REGEXP = $cliArgs['filter'] ?: getenv('TABLE_NAME_REGEXP') ?: ($configFile['export']['table_filter'] ?? '/.*/');

// Output settings
$OUTPUT_DIR = $cliArgs['output'] ?: getenv('OUTPUT_DIR') ?: ($configFile['export']['output_dir'] ?? (__DIR__ . '/export'));
$OUTPUT_FILE = $OUTPUT_DIR . "/schema_{$DB_NAME}.json";
$SCHEMA_VERSION = getenv('SCHEMA_VERSION') ?: date('Y-m-d');

// === SETUP ===
if (!is_dir($OUTPUT_DIR)) {
    mkdir($OUTPUT_DIR, 0777, true);
}

$dsn = "mysql:host=$DB_HOST;dbname=$DB_NAME;charset=utf8mb4";
$options = [
    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
];

// Validate required configuration
if (empty($DB_NAME) || empty($DB_USER)) {
    echo "Error: Database name and user are required." . PHP_EOL;
    echo "Use --database and --user arguments, or set DB_NAME and DB_USER environment variables." . PHP_EOL;
    echo "Run 'php export-schema.php --help' for usage information." . PHP_EOL;
    exit(1);
}

// Validate table filter regex
if (!empty($TABLE_NAME_REGEXP)) {
    if (@preg_match($TABLE_NAME_REGEXP, '') === false) {
        echo "Error: Invalid regex pattern in table filter: {$TABLE_NAME_REGEXP}" . PHP_EOL;
        exit(1);
    }
}

try {
    $pdo = new PDO($dsn, $DB_USER, $DB_PASS, $options);
    echo "Connected to database: {$DB_NAME}@{$DB_HOST}" . PHP_EOL;
} catch (PDOException $e) {
    die("Connection failed: " . $e->getMessage() . PHP_EOL);
}

// Prepare statements (reuse for speed)
$stmtColumns = $pdo->prepare("
    SELECT COLUMN_NAME, COLUMN_TYPE, IS_NULLABLE, COLUMN_DEFAULT, EXTRA, COLUMN_COMMENT, ORDINAL_POSITION
    FROM INFORMATION_SCHEMA.COLUMNS
    WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
    ORDER BY ORDINAL_POSITION
");

$stmtConstraints = $pdo->prepare("
    SELECT tc.CONSTRAINT_NAME, tc.CONSTRAINT_TYPE, kcu.COLUMN_NAME, kcu.ORDINAL_POSITION
    FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
    JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
      ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
     AND tc.TABLE_SCHEMA   = kcu.TABLE_SCHEMA
     AND tc.TABLE_NAME     = kcu.TABLE_NAME
    WHERE tc.TABLE_SCHEMA = ? AND tc.TABLE_NAME = ?
    ORDER BY tc.CONSTRAINT_NAME, kcu.ORDINAL_POSITION
");

$stmtIndexes = $pdo->prepare("
    SELECT INDEX_NAME, NON_UNIQUE, SEQ_IN_INDEX, COLUMN_NAME
    FROM INFORMATION_SCHEMA.STATISTICS
    WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
    ORDER BY INDEX_NAME, SEQ_IN_INDEX
");

$stmtFKs = $pdo->prepare("
    SELECT kcu.CONSTRAINT_NAME, kcu.COLUMN_NAME,
           kcu.REFERENCED_TABLE_NAME AS ref_table,
           kcu.REFERENCED_COLUMN_NAME AS ref_column,
           rc.UPDATE_RULE, rc.DELETE_RULE
    FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
    JOIN INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS rc
      ON rc.CONSTRAINT_NAME  = kcu.CONSTRAINT_NAME
     AND rc.CONSTRAINT_SCHEMA = kcu.TABLE_SCHEMA
    WHERE kcu.TABLE_SCHEMA = ? AND kcu.TABLE_NAME = ?
      AND kcu.REFERENCED_TABLE_NAME IS NOT NULL
");

$stmtInfo = $pdo->prepare("
    SELECT TABLE_ROWS, ENGINE, TABLE_COLLATION
    FROM INFORMATION_SCHEMA.TABLES
    WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
");

// === GET TABLES ===
$tables = $pdo->query("SHOW FULL TABLES WHERE Table_type = 'BASE TABLE'")->fetchAll(PDO::FETCH_NUM);

$tablesOut = [];
$totalTables = count($tables);
$currentTable = 0;

echo "Processing $totalTables tables with filter: {$TABLE_NAME_REGEXP}..." . PHP_EOL;

foreach ($tables as $t) {
    $table = $t[0];
    
    // Skip tables that don't match the regex filter
    if (!preg_match($TABLE_NAME_REGEXP, $table)) {
        continue;
    }
    
    $currentTable++;
    
    echo "[$currentTable/$totalTables] Processing table: $table" . PHP_EOL;

    // Columns
    $stmtColumns->execute([$DB_NAME, $table]);
    $colsRaw = $stmtColumns->fetchAll();

    $columns = array_map(function ($c) {
        return [
            'name'      => $c['COLUMN_NAME'],
            'type'      => $c['COLUMN_TYPE'],
            'nullable'  => ($c['IS_NULLABLE'] === 'YES'),
            'default'   => $c['COLUMN_DEFAULT'],
            'extra'     => $c['EXTRA'],
            'comment'   => $c['COLUMN_COMMENT'],
            'position'  => (int)$c['ORDINAL_POSITION'],
        ];
    }, $colsRaw);

    // Constraints (PRIMARY KEY, UNIQUE, etc.)
    $stmtConstraints->execute([$DB_NAME, $table]);
    $consRaw = $stmtConstraints->fetchAll();
    $primaryKey = [];
    $uniqueConstraints = []; // name => [columns...]

    foreach ($consRaw as $row) {
        if ($row['CONSTRAINT_TYPE'] === 'PRIMARY KEY') {
            $primaryKey[(int)$row['ORDINAL_POSITION']] = $row['COLUMN_NAME'];
        } elseif ($row['CONSTRAINT_TYPE'] === 'UNIQUE') {
            $uniqueConstraints[$row['CONSTRAINT_NAME']][(int)$row['ORDINAL_POSITION']] = $row['COLUMN_NAME'];
        }
    }
    ksort($primaryKey);
    $primaryKey = array_values($primaryKey);

    $uniqueOut = [];
    foreach ($uniqueConstraints as $name => $cols) {
        ksort($cols);
        $uniqueOut[] = ['name' => $name, 'columns' => array_values($cols)];
    }

    // Indexes (including non-unique)
    $stmtIndexes->execute([$DB_NAME, $table]);
    $idxRaw = $stmtIndexes->fetchAll();

    $indexBuckets = []; // name => ['unique'=>bool, 'columns'=>seq=>name]
    foreach ($idxRaw as $r) {
        $name = $r['INDEX_NAME'];
        if (!isset($indexBuckets[$name])) {
            $indexBuckets[$name] = [
                'name'   => $name,
                'unique' => ($r['NON_UNIQUE'] == 0),
                'columns'=> []
            ];
        }
        $indexBuckets[$name]['columns'][(int)$r['SEQ_IN_INDEX']] = $r['COLUMN_NAME'];
    }

    // Flatten/ordered indexes, skip implicit PRIMARY (we already have primaryKey)
    $indexes = [];
    foreach ($indexBuckets as $name => $data) {
        if (strtoupper($name) === 'PRIMARY') continue;
        ksort($data['columns']);
        $indexes[] = [
            'name'    => $name,
            'unique'  => $data['unique'],
            'columns' => array_values($data['columns'])
        ];
    }

    // Foreign Keys
    $stmtFKs->execute([$DB_NAME, $table]);
    $fkRaw = $stmtFKs->fetchAll();

    $fkOut = [];
    // group by constraint name to handle multi-column FKs
    $fkBuckets = [];
    foreach ($fkRaw as $r) {
        $cn = $r['CONSTRAINT_NAME'];
        if (!isset($fkBuckets[$cn])) {
            $fkBuckets[$cn] = [
                'name'        => $cn,
                'columns'     => [],
                'ref_table'   => $r['ref_table'],
                'ref_columns' => [],
                'on_update'   => $r['UPDATE_RULE'],
                'on_delete'   => $r['DELETE_RULE'],
            ];
        }
        $fkBuckets[$cn]['columns'][]     = $r['COLUMN_NAME'];
        $fkBuckets[$cn]['ref_columns'][] = $r['ref_column'];
    }
    $fkOut = array_values($fkBuckets);

    // Table info
    $stmtInfo->execute([$DB_NAME, $table]);
    $info = $stmtInfo->fetch() ?: ['TABLE_ROWS'=>null,'ENGINE'=>null,'TABLE_COLLATION'=>null];

    // DDL
    $ddl = '';
    $ddlStmt = $pdo->query("SHOW CREATE TABLE `{$table}`");
    if ($ddlStmt) {
        $row = $ddlStmt->fetch(PDO::FETCH_ASSOC);
        // MySQL returns keys 'Table' and 'Create Table'; numeric index 1 also works
        $ddl = $row['Create Table'] ?? ($row[1] ?? '');
    }

    $tablesOut[] = [
        'name'         => $table,
        'columns'      => $columns,
        'primary_key'  => $primaryKey,
        'unique'       => $uniqueOut,
        'indexes'      => $indexes,
        'foreign_keys' => $fkOut,
        'table_info'   => [
            'row_count_est' => isset($info['TABLE_ROWS']) ? (int)$info['TABLE_ROWS'] : null,
            'engine'        => $info['ENGINE'] ?? null,
            'collation'     => $info['TABLE_COLLATION'] ?? null,
        ],
        'ddl'          => $ddl,
    ];
}

// === PACKAGE & WRITE ===
$out = [
    'database'       => $DB_NAME,
    'generated_at'   => date('c'),
    'schema_version' => $SCHEMA_VERSION,
    'tables'         => $tablesOut,
];

$jsonOutput = json_encode($out, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);

if ($jsonOutput === false) {
    echo "Error: Failed to encode JSON data" . PHP_EOL;
    exit(1);
}

if (file_put_contents($OUTPUT_FILE, $jsonOutput) === false) {
    echo "Error: Failed to write output file: $OUTPUT_FILE" . PHP_EOL;
    exit(1);
}

$filteredCount = count($tablesOut);
$fileSize = number_format(filesize($OUTPUT_FILE) / 1024, 1);
echo "Export complete: $OUTPUT_FILE" . PHP_EOL;
echo "Exported $filteredCount tables, {$fileSize} KB" . PHP_EOL;

File: export-schema.py
#!/usr/bin/env python3
"""
MySQL Schema Export Utility

Exports MySQL/MariaDB database schema to JSON format with comprehensive metadata
including columns, constraints, indexes, foreign keys, and DDL statements.

@author Bill Nobes
@license MIT
@version 1.0.0
"""

import argparse
import configparser
import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path

import pymysql.cursors


def show_help():
    """Display help information"""
    print("MySQL Schema Export Utility\n")
    print("Usage: python export-schema.py [OPTIONS]\n")
    print("Options:")
    print("  --host HOST          Database host (default: localhost)")
    print("  --database DB        Database name (required)")
    print("  --user USER          Database user (required)")
    print("  --password PASS      Database password")
    print("  --port PORT          Database port (default: 3306)")
    print("  --output DIR         Output directory (default: ./export)")
    print("  --filter REGEX       Table name filter regex (default: .* for all tables)")
    print("  --help, -h           Show this help message\n")
    print("Environment Variables:")
    print("  DB_HOST, DB_NAME, DB_USER, DB_PASS, DB_PORT")
    print("  OUTPUT_DIR, TABLE_NAME_REGEXP, SCHEMA_VERSION\n")
    print("Examples:")
    print("  python export-schema.py --database mydb --user root")
    print("  python export-schema.py --db mydb --user root --filter 'user_.*'")
    print("  DB_NAME=mydb DB_USER=root python export-schema.py")


def load_config_file(filename='config.ini'):
    """Load configuration from INI file"""
    if not os.path.exists(filename):
        return {}
    
    config = configparser.ConfigParser()
    try:
        config.read(filename)
        return config
    except Exception as e:
        print(f"Warning: Could not parse config file {filename}: {e}")
        return {}


def parse_args():
    """Parse command line arguments"""
    parser = argparse.ArgumentParser(description='MySQL Schema Export Utility', add_help=False)
    parser.add_argument('--help', '-h', action='store_true', help='Show help message')
    parser.add_argument('--host', help='Database host')
    parser.add_argument('--database', '--db', help='Database name')
    parser.add_argument('--user', help='Database user')
    parser.add_argument('--password', help='Database password')
    parser.add_argument('--port', type=int, help='Database port')
    parser.add_argument('--output', help='Output directory')
    parser.add_argument('--filter', help='Table name filter regex')
    
    return parser.parse_args()


def get_config_value(cli_arg, env_var, config_file, section, key, default=None):
    """Get configuration value with priority: CLI > env > config file > default"""
    if cli_arg is not None:
        return cli_arg
    
    env_val = os.getenv(env_var)
    if env_val is not None:
        return env_val
    
    if config_file and section in config_file and key in config_file[section]:
        return config_file[section][key]
    
    return default


def get_table_columns(connection, db_name, table_name):
    """Get column information for a table"""
    query = """
        SELECT COLUMN_NAME, COLUMN_TYPE, IS_NULLABLE, COLUMN_DEFAULT, 
               EXTRA, COLUMN_COMMENT, ORDINAL_POSITION
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s
        ORDER BY ORDINAL_POSITION
    """
    
    with connection.cursor() as cursor:
        cursor.execute(query, (db_name, table_name))
        rows = cursor.fetchall()
        
        columns = []
        for row in rows:
            columns.append({
                'name': row['COLUMN_NAME'],
                'type': row['COLUMN_TYPE'],
                'nullable': row['IS_NULLABLE'] == 'YES',
                'default': row['COLUMN_DEFAULT'],
                'extra': row['EXTRA'],
                'comment': row['COLUMN_COMMENT'],
                'position': int(row['ORDINAL_POSITION'])
            })
        
        return columns


def get_table_constraints(connection, db_name, table_name):
    """Get constraint information for a table"""
    query = """
        SELECT tc.CONSTRAINT_NAME, tc.CONSTRAINT_TYPE, kcu.COLUMN_NAME, kcu.ORDINAL_POSITION
        FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
        JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
          ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
         AND tc.TABLE_SCHEMA = kcu.TABLE_SCHEMA
         AND tc.TABLE_NAME = kcu.TABLE_NAME
        WHERE tc.TABLE_SCHEMA = %s AND tc.TABLE_NAME = %s
        ORDER BY tc.CONSTRAINT_NAME, kcu.ORDINAL_POSITION
    """
    
    with connection.cursor() as cursor:
        cursor.execute(query, (db_name, table_name))
        rows = cursor.fetchall()
        
        primary_key = {}
        unique_constraints = {}
        
        for row in rows:
            if row['CONSTRAINT_TYPE'] == 'PRIMARY KEY':
                primary_key[int(row['ORDINAL_POSITION'])] = row['COLUMN_NAME']
            elif row['CONSTRAINT_TYPE'] == 'UNIQUE':
                constraint_name = row['CONSTRAINT_NAME']
                if constraint_name not in unique_constraints:
                    unique_constraints[constraint_name] = {}
                unique_constraints[constraint_name][int(row['ORDINAL_POSITION'])] = row['COLUMN_NAME']
        
        # Sort and flatten primary key
        sorted_pk = [primary_key[pos] for pos in sorted(primary_key.keys())]
        
        # Sort and flatten unique constraints
        unique_out = []
        for name, cols in unique_constraints.items():
            sorted_cols = [cols[pos] for pos in sorted(cols.keys())]
            unique_out.append({'name': name, 'columns': sorted_cols})
        
        return sorted_pk, unique_out


def get_table_indexes(connection, db_name, table_name):
    """Get index information for a table"""
    query = """
        SELECT INDEX_NAME, NON_UNIQUE, SEQ_IN_INDEX, COLUMN_NAME
        FROM INFORMATION_SCHEMA.STATISTICS
        WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s
        ORDER BY INDEX_NAME, SEQ_IN_INDEX
    """
    
    with connection.cursor() as cursor:
        cursor.execute(query, (db_name, table_name))
        rows = cursor.fetchall()
        
        index_buckets = {}
        
        for row in rows:
            index_name = row['INDEX_NAME']
            if index_name not in index_buckets:
                index_buckets[index_name] = {
                    'name': index_name,
                    'unique': row['NON_UNIQUE'] == 0,
                    'columns': {}
                }
            index_buckets[index_name]['columns'][int(row['SEQ_IN_INDEX'])] = row['COLUMN_NAME']
        
        # Skip PRIMARY index and flatten columns
        indexes = []
        for name, data in index_buckets.items():
            if name.upper() == 'PRIMARY':
                continue
            sorted_cols = [data['columns'][pos] for pos in sorted(data['columns'].keys())]
            indexes.append({
                'name': name,
                'unique': data['unique'],
                'columns': sorted_cols
            })
        
        return indexes


def get_table_foreign_keys(connection, db_name, table_name):
    """Get foreign key information for a table"""
    query = """
        SELECT kcu.CONSTRAINT_NAME, kcu.COLUMN_NAME,
               kcu.REFERENCED_TABLE_NAME AS ref_table,
               kcu.REFERENCED_COLUMN_NAME AS ref_column,
               rc.UPDATE_RULE, rc.DELETE_RULE
        FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
        JOIN INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS rc
          ON rc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
         AND rc.CONSTRAINT_SCHEMA = kcu.TABLE_SCHEMA
        WHERE kcu.TABLE_SCHEMA = %s AND kcu.TABLE_NAME = %s
          AND kcu.REFERENCED_TABLE_NAME IS NOT NULL
    """
    
    with connection.cursor() as cursor:
        cursor.execute(query, (db_name, table_name))
        rows = cursor.fetchall()
        
        fk_buckets = {}
        
        for row in rows:
            constraint_name = row['CONSTRAINT_NAME']
            if constraint_name not in fk_buckets:
                fk_buckets[constraint_name] = {
                    'name': constraint_name,
                    'columns': [],
                    'ref_table': row['ref_table'],
                    'ref_columns': [],
                    'on_update': row['UPDATE_RULE'],
                    'on_delete': row['DELETE_RULE']
                }
            fk_buckets[constraint_name]['columns'].append(row['COLUMN_NAME'])
            fk_buckets[constraint_name]['ref_columns'].append(row['ref_column'])
        
        return list(fk_buckets.values())


def get_table_info(connection, db_name, table_name):
    """Get table metadata information"""
    query = """
        SELECT TABLE_ROWS, ENGINE, TABLE_COLLATION
        FROM INFORMATION_SCHEMA.TABLES
        WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s
    """
    
    with connection.cursor() as cursor:
        cursor.execute(query, (db_name, table_name))
        row = cursor.fetchone()
        
        if row:
            return {
                'row_count_est': int(row['TABLE_ROWS']) if row['TABLE_ROWS'] is not None else None,
                'engine': row['ENGINE'],
                'collation': row['TABLE_COLLATION']
            }
        
        return {'row_count_est': None, 'engine': None, 'collation': None}


def get_table_ddl(connection, table_name):
    """Get CREATE TABLE statement for a table"""
    query = f"SHOW CREATE TABLE `{table_name}`"
    
    with connection.cursor() as cursor:
        cursor.execute(query)
        row = cursor.fetchone()
        
        if row:
            # The result has keys 'Table' and 'Create Table'
            return row.get('Create Table', '')
        
        return ''


def main():
    """Main execution function"""
    # Parse arguments and load configuration
    args = parse_args()
    
    if args.help:
        show_help()
        sys.exit(0)
    
    config_file = load_config_file()
    
    # Get configuration values with priority: CLI > env > config file > defaults
    db_host = get_config_value(args.host, 'DB_HOST', config_file, 'database', 'host', 'localhost')
    db_name = get_config_value(args.database, 'DB_NAME', config_file, 'database', 'name', '')
    db_user = get_config_value(args.user, 'DB_USER', config_file, 'database', 'user', '')
    db_pass = get_config_value(args.password, 'DB_PASS', config_file, 'database', 'password', '')
    db_port = int(get_config_value(args.port, 'DB_PORT', config_file, 'database', 'port', 3306))
    
    table_filter = get_config_value(args.filter, 'TABLE_NAME_REGEXP', config_file, 'export', 'table_filter', '.*')
    output_dir = get_config_value(args.output, 'OUTPUT_DIR', config_file, 'export', 'output_dir', './export')
    schema_version = os.getenv('SCHEMA_VERSION', datetime.now().strftime('%Y-%m-%d'))
    
    # Convert PHP-style regex delimiters to Python regex if needed
    if table_filter.startswith('/') and table_filter.endswith('/'):
        table_filter = table_filter[1:-1]  # Remove PHP delimiters
    
    # Validate required configuration
    if not db_name or not db_user:
        print("Error: Database name and user are required.")
        print("Use --database and --user arguments, or set DB_NAME and DB_USER environment variables.")
        print("Run 'python export-schema.py --help' for usage information.")
        sys.exit(1)
    
    # Validate table filter regex
    try:
        re.compile(table_filter)
    except re.error as e:
        print(f"Error: Invalid regex pattern in table filter: {table_filter} - {e}")
        sys.exit(1)
    
    # Create output directory
    Path(output_dir).mkdir(parents=True, exist_ok=True)
    output_file = os.path.join(output_dir, f"schema_{db_name}.json")
    
    # Connect to database
    try:
        connection = pymysql.connect(
            host=db_host,
            port=db_port,
            user=db_user,
            password=db_pass,
            database=db_name,
            charset='utf8mb4',
            cursorclass=pymysql.cursors.DictCursor
        )
        print(f"Connected to database: {db_name}@{db_host}")
    except Exception as e:
        print(f"Connection failed: {e}")
        sys.exit(1)
    
    try:
        # Get all tables
        with connection.cursor() as cursor:
            cursor.execute("SHOW FULL TABLES WHERE Table_type = 'BASE TABLE'")
            all_tables = [row[f'Tables_in_{db_name}'] for row in cursor.fetchall()]
        
        total_tables = len(all_tables)
        current_table = 0
        tables_out = []
        
        print(f"Processing {total_tables} tables with filter: {table_filter}...")
        
        # Filter and process tables
        pattern = re.compile(table_filter)
        
        for table in all_tables:
            # Skip tables that don't match the regex filter
            if not pattern.match(table):
                continue
            
            current_table += 1
            print(f"[{current_table}/{total_tables}] Processing table: {table}")
            
            # Get table components
            columns = get_table_columns(connection, db_name, table)
            primary_key, unique_constraints = get_table_constraints(connection, db_name, table)
            indexes = get_table_indexes(connection, db_name, table)
            foreign_keys = get_table_foreign_keys(connection, db_name, table)
            table_info = get_table_info(connection, db_name, table)
            ddl = get_table_ddl(connection, table)
            
            # Add to output
            tables_out.append({
                'name': table,
                'columns': columns,
                'primary_key': primary_key,
                'unique': unique_constraints,
                'indexes': indexes,
                'foreign_keys': foreign_keys,
                'table_info': table_info,
                'ddl': ddl
            })
        
        # Create output structure
        output = {
            'database': db_name,
            'generated_at': datetime.now().isoformat(),
            'schema_version': schema_version,
            'tables': tables_out
        }
        
        # Write JSON file
        try:
            with open(output_file, 'w', encoding='utf-8') as f:
                json.dump(output, f, indent=2, ensure_ascii=False)
            
            filtered_count = len(tables_out)
            file_size = os.path.getsize(output_file) / 1024
            print(f"Export complete: {output_file}")
            print(f"Exported {filtered_count} tables, {file_size:.1f} KB")
            
        except Exception as e:
            print(f"Error: Failed to write output file {output_file}: {e}")
            sys.exit(1)
    
    finally:
        connection.close()


if __name__ == '__main__':
    main()